pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -856,37 +856,19 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 			// On failure, return error. Caller will log and retry.
 			return volumeToMount.GenerateError("MapVolume.SetUp failed", mapErr)
 		}
-		// Update devicePath for none attachable plugin case
-		if len(devicePath) == 0 {
-			if len(pluginDevicePath) != 0 {
-				devicePath = pluginDevicePath
-			} else {
-				return volumeToMount.GenerateError("MapVolume failed", fmt.Errorf("Device path of the volume is empty"))
-			}
-		}
-		// Update actual state of world to reflect volume is globally mounted
-		markDeviceMappedErr := actualStateOfWorld.MarkDeviceAsMounted(
-			volumeToMount.VolumeName, devicePath, globalMapPath)
-		if markDeviceMappedErr != nil {
-			// On failure, return error. Caller will log and retry.
-			return volumeToMount.GenerateError("MapVolume.MarkDeviceAsMounted failed", markDeviceMappedErr)
-		}
 
-		mapErr = og.blkUtil.MapDevice(devicePath, globalMapPath, string(volumeToMount.Pod.UID))
-		if mapErr != nil {
-			// On failure, return error. Caller will log and retry.
-			return volumeToMount.GenerateError("MapVolume.MapDevice failed", mapErr)
+		// if pluginDevicePath is provided, assume attacher may not provide device
+		// or attachment flow uses SetupDevice to get device path
+		if len(pluginDevicePath) != 0 {
+			devicePath = pluginDevicePath
+		}
+		if len(devicePath) == 0 {
+			return volumeToMount.GenerateError("MapVolume failed", fmt.Errorf("Device path of the volume is empty"))
```

Comment:
`devicePath` comes from `volumeAttacher.WaitForAttach(...)` or `blockVolumeMapper.SetUpDevice()`. Is that correct? Will all plugins return `devicePath` in one of those functions? Will CSI?

Reply:
The original code assumes that devicePath comes from either a `volumeAttacher.WaitForAttach(...)` or from `blockVolumeMapper.SetupDevice()` .  If neither one return a path, code errors out.  So yes, it is a assume that a `blockVolumeMapper` will return a deviceMap.

Since CSI does not return a devicePath with `WaitForAttach`, it uses `SetupDevice` to return the global map path (which is the path sent to the external driver for NodeStage call).

Reply:
Ack!